### PR TITLE
Feature/simd vec3

### PR DESCRIPTION
**Pull request by andrearastelli**
Following #64

### CODE REVIEW

**Comment by andrearastelli** [comment]
@Ale32 all Vec3 is now SIMD-ed.

It's time for testing it!!!!

**Comment by andrearastelli** [comment]
@Ale32 Added GoogleTests package as a dependency (`ext` folder).
This way we can test any update in the system that may require some big change.

The `CMakeLists.txt` file required some updates as well, but cannot understand how the testing system can work inside the `CMake` execution.
Need to dig deeper into that.

Right now, in `CLion` I can run the tests separately from the main target, but this doesn't solve the test automation process.
Next step - will add the test automation into the `CMakeLists.txt` file so every time we send the project to travis or appvejor the testing will occur by default.

Updated `Vec3` using a union to share the memory representation used by the `float v_f[4]` and the `__m128 v` object. This way we don't need to worry about converting back and forth the data from the SIMD-like representation and the usable `float` representation.

Current Issue: if I try to render the image, I get black pixels.
Need to analyse that in more detail.
As a start I'm currently testing all the Vec3 class using gtest - because that is the only class that is changed - and try to figure out where the problem might be, in the conversion process.

**Comment by andrearastelli** [comment]
@Ale32 Rebased to be on top of `develop`

**Comment by Ale32** [comment]
@andrearastelli cool! Thanks for your effort. Using google test is a good idea! I'll check what I get on my windows machine.

**Comment by andrearastelli** [comment]
@Ale32 I've spent some time trying to figure out the linux/windows SIMD setup, but this require a more thorough test.

The gtest are correctly setup, and at this point we can safely test against the current Vec3 or the SIMD version, and with some confidence expect the same results.

**Comment by andrearastelli** [comment]
@Ale32 before merging this, I need to figure a way to test SIMD in both travis and appveyor.
The issue with travis is that the function `_mm_hadd_ps(__m128, __m128)` is not recognized, for some reason (maybe the SIMD in travis is `SSE` and this function is available in `SSE2` or something like that).
The issues with appveyor are so many that listing those here doesn't make any sense, and I need to properly access Window and start developing there at this point, so I can commit from there the working version for appveyor.

**Comment by Ale32** [comment]
I'm not able to compile the code on Windows. I guess the issue is something related to this https://stackoverflow.com/questions/11228855/header-files-for-x86-simd-intrinsics#

**Comment by andrearastelli** [comment]
@Ale32 ...no the issue was simpler than that...
```git submodule update --init --recursive```
If you fetch the last update from this branch you'll be able to compile everything.

The only issue I got with windows is that I cannot run the gTest suite, for some reason that is not exactly clear to me.. So, for the time being, in this branch, the `CMakeLists.txt` file has the test building disabled for windows, and will remain this way until we figure a solution to this problem.

File: src/vec3.h [hunk 1]
```diff
@@ -5,12 +5,20 @@
...
 #include <cmath>
 
+#if defined(_MSC_VER)
+#include <immintrin.h>
+#else
+#include <x86intrin.h>
+#endif
 
-class Vec3
+class alignas(16) Vec3
```
Comment (andrearastelli) [inline src/vec3.h]: @Ale32 The previous definition for memory aligned class arguments was very old and not supported anymore.
This one should work for windows as well.

**Comment by andrearastelli** [comment]
@Ale32 Color is SIMD-ed as well. This let us gain another 2% of performance.

Now some proper profiling is required to avoid random number in performance improvements.

**Comment by andrearastelli** [comment]
Task required in order to properly continue working on this PR:
- [ ] Make the tests work on `travis` and `appveyor`.
- [ ] Make the `gTest` suite work on both Linux and Windows.
- [ ] Profile the SIMD code against the old one, to have a proper performance table.

**Comment by andrearastelli** [comment]
@Ale32 Fixed travis testing setup. Now I need to figure how to not execute the `_mm_hadd_ps` when the hardware doesn't support it.

**Comment by andrearastelli** [comment]
This PR is starting to be **huge**, compared to the other ones.

Anyway.. the stuff that is updated, so far:
# Vec3 now runs with parallel code
# Color now runs with parallel code
# `simd.h` added. Because some of the features of Vec3 and Color requires horizontal sums, and only starting from SSE3 and forward those implementation actually take place, like `_mm_hadd_ps`, a new SIMD check system needed to be designed
# For reference about SIMD on intel platform: https://software.intel.com/sites/landingpage/IntrinsicsGuide/#expand=2777,2777&text=_mm_hadd_ps
# Various performance and memory check helped me figure that the main function could have run with far less memory allocated - so fixed that
# To test the SIMD implementation the `gTest` package needed to be added
# Various test implemented for `Vec3`
# Various test implemented for `Color`
# The `.travis.yml` script required a big update to support test cases
# The `appveyor.yml` script required some updates to support test cases (but still have lots of issues)

Stuff that is still not finished:
- [ ] Have `appveyor.yml` working and testing the proper code on windows
- [ ] Have the windows implementation work the same as the Linux one

@Ale32 This is the current state of the art of this PR.

**Comment by andrearastelli** [comment]
@Ale32 ALL TEST PASSED SUCCESSFULLY!!!!!!!

**Comment by andrearastelli** [comment]
Rebased to be on top of develop.

File: src/color.h [hunk 2]
```diff
@@ -38,7 +47,7 @@ class Color
      * @param g Green component.
      * @param b Blue component.
      */
-    Color(int r, int g, int b): c{r / 255.0f, g / 255.0f, b / 255.0f} {}
+    Color(int r, int g, int b, int alpha=255): v{_mm_set_ps(alpha/255.0f, b/255.0f, g/255.0f, r/255.0f)} {}
```
Comment (Ale32) [inline src/color.h]: Why (alpha, b, g, r)?

File: src/hitable.h [hunk 3]
```diff
@@ -132,6 +132,7 @@ RotateY::RotateY(Hitable *p, float angle) : ptr(p)
         for (auto j=0; j<2; ++j)
             for (auto k=0; k<2; ++k)
             {
+                // TOODO: this can be SIMDed - needs more check.
```
Comment (Ale32) [inline src/hitable.h]: Typo?

**Pull request merged**